src/main/scala/util/CreditedIO.scala

```diff
@@ -0,0 +1,152 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util
+
+import chisel3._
+import chisel3.util._
+
+/** Transmission delay in credit-debit systems.
+  * debit delay is the number of cycles it takes debit+'bits' to transit the link.
+  * credit delay is the number of cycles it takes credits to be returned.
+  * round trip / total delay is the sum of debit and credit delay.
+  * The system must have a positive total delay, otherwise you have a combinational loop.
+  */
+case class CreditedDelay(debit: Int, credit: Int)
+{
+  val total = debit + credit
+  require (debit >= 0)
+  require (credit >= 0)
+
+  def flip: CreditedDelay = CreditedDelay(credit, debit)
+
+  def +(that: CreditedDelay): CreditedDelay =
+    CreditedDelay(debit + that.debit, credit + that.credit)
+
+  override def toString = s"${debit}:${credit}"
+}
+
+/** CreditedIO provides credit-debit-based flow control for arbitrary Data.
+  * The sender may only transmit Data when it has non-zero credits.
+  * Receivers provide 0 or 1 credits to senders using the credit field.
+  * Senders consume 0 or 1 credits by setting the debit field.
+  * The bits Data field is DontCare when debit=0.
+  * credit MAY depend combinationally on debit.
+  * debit MAY depend combinationally on credit.
+  * WARNING: The user must ensure the round trip time is > 0.
+  * Failure to comply will result in a combinational loop!
+  */
+final class CreditedIO[T <: Data](gen: T) extends Bundle
+{
+  override def cloneType: this.type = new CreditedIO(genType).asInstanceOf[this.type]
+  def genType: T = gen
+
+  val credit = Input (Bool()) // 1: a credit is given to the sender by the receiver
+  val debit  = Output(Bool()) // 1: a credit is consumed by the sender to transfer 'bits'
+  val bits   = Output(genType)
+
+  /** Provide a DecoupledIO interface for sending CreditedIO[Data].
+    * Convert an IrrevocableIO input to DecoupledIO via Decoupled().
```

Comment:
`IrrevocableIO`?

Reply:
What are you asking? You want me to remove that line of commentary?
While the sender function returns a DecoupledIO, you may want to connect an IrrevocableIO.

Reply:
I just didn't see `IrrevocableIO` mentioned anywhere in this particular function. `this` is `CreditedIO` and as you the say the return type is `DecoupledIO`. I guess I would find "To connect an IrrevocableIO sender, first convert it with `Decoupled()`" to be clearer phrasing to me?